hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -649,3 +649,45 @@ def nested_dtypes(
     return st.recursive(
         subtype_strategy, lambda x: array_dtypes(x, allow_subarrays=True), max_leaves
     ).filter(lambda d: max_itemsize is None or d.itemsize <= max_itemsize)
+
+
+@st.defines_strategy
+def valid_tuple_axes(ndim, min_size=0, max_size=None):
+    # type: (int, int, Optional[int]) -> st.SearchStrategy[Tuple[int, ...]]
+    """Return a strategy for generating permissible tuple-values for the
+    ``axis`` argument for a numpy sequential function (e.g.
+    :func:`numpy.sum <numpy:numpy.sum>`), given an array of the specified
+    dimensionality.
+
+    All tuples will have an length >= min_size and <= max_size. The default
+    value for max_len is ``ndim``.
+
+    Examples from this strategy shrink towards an empty tuple.
+
+
+    .. code-block:: pycon
+
+        >>> [valid_tuple_axes(3).example() for i in range(4)]
+        [(-3, 1), (0, 1, -1), (0, 2), (0, -2, 2)]
+    """
+    if max_size is None:
+        max_size = ndim
+
+    check_type(integer_types, ndim, "ndim")
+    check_valid_magnitude(ndim, "ndim")
+
+    order_check("len", 0, min_size, max_size)
+    check_argument(
+        min_size <= ndim, "min_len={} is larger than ndim={}", min_size, ndim
+    )
+    check_type(integer_types, min_size, "min_len")
+    check_type(integer_types, max_size, "max_len")
+
+    max_size = min(ndim, max_size)
+
+    # shrink axis values from negative to positive
+    return st.integers(min_size, max_size).flatmap(
```

Comment:
Ah, yeah this is much nicer. The only issue is that this does not reliably shrink towards positive axis-values, which I feel is an important property for this strategy. 

E.g. suppose `ndim = 3` and one of my test fails on `axis -> (0, 1, 2)`, this version of the strategy will often shrink this case to `(0, 1, -1)` or `(0, -2, -1)`, which would lead me to believe that something subtle is afoot with outer axes. Shrinking to `(0, 1, 2)` tells a much more plain story.

What do you think of:
```python
    return st.lists(
    st.integers(0, max(2*ndim - 1, 0)),
        min_size=min_size,
        max_size=max_size,
        unique_by=lambda x: x if x < ndim else x - ndim,
    ).map(lambda x: tuple(i - 2*ndim if ndim <= i else i for i in x))
```

which ultimately maps the drawn interval `[0, 2*ndim - 1]` back to `[-ndim, ndim - 1]`, and shrinks each axis index towards its positive representation.

Reply:
Also, thanks for linking to #1790 ! I have only been able to read part of it thus far, but it looks to be  **super** useful! 🙌

Reply:
Looks good to me, though I'd do it with an inner `.map` on the axis numbers like so:

```python
axes = st.integers(0, max(0, 2 * ndim - 1)).map(lambda x: x if x < ndim else x - 2 * ndim)
return st.lists(axes, min_size, max_size, unique_by=lambda x: x % ndim).map(tuple)
```

We're basically in the realm of unsatisfying trade-offs here though, because none of these will shrink efficiently subject to the constraint that `len(axis) == ndim`.  Using `integers()` directly lets the engine flip a sign bit but doesn't find the all-positive minimal example, so I do prefer this option overall.

And I'm glad the shrinking guide is useful!  I would love any comments - you have a very different perspective on this to David and are much closer to the target audience 😄